src/coprocessor/dag/batch/executors/util/heap.rs

```diff
@@ -0,0 +1,251 @@
+// Copyright 2017 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::cmp::{self, Ordering};
+use std::collections::BinaryHeap;
+use std::sync::Arc;
+use std::usize;
+
+use crate::coprocessor::codec::batch::LazyBatchColumnVec;
+use crate::coprocessor::dag::expr::Result;
+
+const HEAP_MAX_CAPACITY: usize = 1024;
+
+/// SortRow wrapping a logical row will be used in the TopNHeap.
+pub struct SortRow {
```

Comment:
How about:

```rust
struct DataView {
    chunks: Vec<LazyBatchColumnVec>,  // Multiple inputs places one by one
    order_columns: Vec<(usize, bool)>,  // (column offset, descending)
}

pub struct HeapRow {
    view: Arc<DataView>,
    row_index: (usize, usize),  // (chunk offset, row offset in chunk)
}
```

In this way, no copies for each batch input.

Reply:
Here I do not use `Vec<LazyBatchColumn>`, because I only want to keep `LazyBatchColumn` which contains row included in the  Heap, for those `LazyBatchColumn` which is not included in the heap, I'd like to throw it once it was not included. Then the worse case is each row in the heap was from a `LazyBatchColumn`, so we will have at most `limit` LazyBatchColumn in memory instead of all `LazyBatchOlumn`

Reply:
Got it.